engines/sherlock/surface.cpp

```diff
@@ -0,0 +1,189 @@
+		addDirtyRect(destRect);
+		_surface.copyRectToSurface(src, destRect.left, destRect.top, srcRect);
+	}
+}
+
+void Surface::blitFrom(const ImageFrame &src, const Common::Point &pt, const Common::Rect &srcBounds) {
+	blitFrom(src._frame, pt, srcBounds);
+}
+
+void Surface::blitFrom(const Surface &src, const Common::Point &pt, const Common::Rect &srcBounds) {
+	blitFrom(src._surface, pt, srcBounds);
+}
+
+void Surface::transBlitFrom(const ImageFrame &src, const Common::Point &pt,
+		bool flipped, int overrideColor) {
+	transBlitFrom(src._frame, pt + src._offset, flipped, overrideColor);
+}
+
+void Surface::transBlitFrom(const Graphics::Surface &src, const Common::Point &pt,
+		bool flipped, int overrideColor) {
+	Common::Rect drawRect(0, 0, src.w, src.h);
+	Common::Rect destRect(pt.x, pt.y, pt.x + src.w, pt.y + src.h);
+
+	// Clip the display area to on-screen
+	if (!clip(drawRect, destRect))
+		// It's completely off-screen
+		return;
+
+	if (flipped)
+		drawRect = Common::Rect(src.w - drawRect.right, src.h - drawRect.bottom,
+			src.w - drawRect.left, src.h - drawRect.top);
+
+	Common::Point destPt(destRect.left, destRect.top);
+	addDirtyRect(Common::Rect(destPt.x, destPt.y, destPt.x + drawRect.width(),
+		destPt.y + drawRect.height()));
+
+	// Draw loop
+	const int TRANSPARENCY = 0xFF;
+	for (int yp = 0; yp < drawRect.height(); ++yp) {
+		const byte *srcP = (const byte *)src.getBasePtr(
+			flipped ? drawRect.right - 1 : drawRect.left, drawRect.top + yp);
+		byte *destP = (byte *)getBasePtr(destPt.x, destPt.y + yp);
+
+		for (int xp = 0; xp < drawRect.width(); ++xp, ++destP) {
+			if (*srcP != TRANSPARENCY)
+				*destP = overrideColor ? overrideColor : *srcP;
+
+			srcP = flipped ? srcP - 1 : srcP + 1;
+		}
+	}
+}
+
+void Surface::fillRect(int x1, int y1, int x2, int y2, byte color) {
+	fillRect(Common::Rect(x1, y1, x2, y2), color);
+}
+
+void Surface::fillRect(const Common::Rect &r, byte color) {
+	_surface.fillRect(r, color);
+	addDirtyRect(r);
+}
+
+bool Surface::clip(Common::Rect &srcBounds, Common::Rect &destBounds) {
+	if (destBounds.left >= _surface.w || destBounds.top >= _surface.h ||
+			destBounds.right <= 0 || destBounds.bottom <= 0)
+		return false;
+
+	// Clip the bounds if necessary to fit on-screen
+	if (destBounds.right > _surface.w) {
+		srcBounds.right -= destBounds.right - _surface.w;
+		destBounds.right = _surface.w;
+	}
+
+	if (destBounds.bottom > _surface.h) {
+		srcBounds.bottom -= destBounds.bottom - _surface.h;
+		destBounds.bottom = _surface.h;
+	}
+
+	if (destBounds.top < 0) {
+		srcBounds.top += -destBounds.top;
+		destBounds.top = 0;
+	}
+
+	if (destBounds.left < 0) {
+		srcBounds.left += -destBounds.left;
+		destBounds.left = 0;
+	}
+
+	return true;
+}
+
+void Surface::clear() {
+	fillRect(Common::Rect(0, 0, _surface.w, _surface.h), 0);
+}
+
+void Surface::free() {
+	if (_freePixels) {
+		_surface.free();
+		_freePixels = false;
+	}
+}
+
+void Surface::setPixels(byte *pixels, int w, int h) {
```

Comment:
GCC warns that 'w' and 'h' shadow the w() and h() methods of the Surface class. I guess that's why other methods use 'width' and 'height' as parameter names.